File: lib/models/labels/social_media_label.dart
import 'dart:io';

/// Social media profile label types.
///
/// | Label          | Android | iOS |
/// |----------------|:-------:|:---:|
/// | aim            | ✔       | ✔   |
/// | facebook       | ⨯       | ✔   |
/// | flickr         | ⨯       | ✔   |
/// | gameCenter     | ⨯       | ✔   |
/// | gaduGadu       | ⨯       | ✔   |
/// | googleTalk     | ✔       | ✔   |
/// | icq            | ✔       | ✔   |
/// | jabber         | ✔       | ✔   |
/// | linkedIn       | ⨯       | ✔   |
/// | msn            | ✔       | ✔   |
/// | mySpace        | ⨯       | ✔   |
/// | netmeeting     | ✔       | ⨯   |
/// | qq             | ✔       | ✔   |
/// | sinaWeibo      | ⨯       | ✔   |
/// | skype          | ✔       | ✔   |
/// | tencentWeibo   | ⨯       | ✔   |
/// | twitter        | ⨯       | ✔   |
/// | yahoo          | ✔       | ✔   |
/// | yelp           | ⨯       | ✔   |
/// | other¹         | ✔       | ✔   |
/// | custom         | ✔       | ✔   |
///
/// ¹ The other label is not natively supported on either platform and will
/// be converted to custom with "other" as the raw string value. It was included
/// to provide a default label that is not arbitrarily one of the supported
/// providers.
enum SocialMediaLabel {
  aim,
  facebook,
  flickr,
  gameCenter,
  gaduGadu,
  googleTalk,
  icq,
  jabber,
  linkedIn,
  msn,
  mySpace,
  netmeeting,
  qq,
  sinaWeibo,
  skype,
  tencentWeibo,
  twitter,
  yahoo,
  yelp,
  other,
  custom;

  static const _android = {
    SocialMediaLabel.aim,
    SocialMediaLabel.googleTalk,
    SocialMediaLabel.icq,
    SocialMediaLabel.jabber,
    SocialMediaLabel.msn,
    SocialMediaLabel.netmeeting,
    SocialMediaLabel.qq,
    SocialMediaLabel.skype,
    SocialMediaLabel.yahoo,
    SocialMediaLabel.other,
    SocialMediaLabel.custom,
  };

  static const _apple = {
    SocialMediaLabel.aim,
    SocialMediaLabel.facebook,
    SocialMediaLabel.flickr,
    SocialMediaLabel.gameCenter,
    SocialMediaLabel.gaduGadu,
    SocialMediaLabel.googleTalk,
    SocialMediaLabel.icq,
    SocialMediaLabel.jabber,
    SocialMediaLabel.linkedIn,
    SocialMediaLabel.msn,
    SocialMediaLabel.mySpace,
    SocialMediaLabel.qq,
    SocialMediaLabel.sinaWeibo,
    SocialMediaLabel.skype,
    SocialMediaLabel.tencentWeibo,
    SocialMediaLabel.twitter,
    SocialMediaLabel.yahoo,
    SocialMediaLabel.yelp,
    SocialMediaLabel.other,
    SocialMediaLabel.custom,
  };

  bool get isSupported {
    if (Platform.isAndroid) return _android.contains(this);
    if (Platform.isIOS || Platform.isMacOS) return _apple.contains(this);
    return true;
  }
}

File: ios/flutter_contacts/Sources/flutter_contacts/crud/models/labels/SocialMediaLabel.swift
import Contacts
import Foundation

enum SocialMediaLabel: String, CaseIterable {
    case aim
    case facebook
    case flickr
    case gameCenter
    case gaduGadu
    case googleTalk
    case icq
    case jabber
    case linkedIn
    case msn
    case mySpace
    case netmeeting
    case qq
    case sinaWeibo
    case skype
    case tencentWeibo
    case twitter
    case yahoo
    case yelp
    case other
    case custom

    private static let instantMessageLabels: Set<SocialMediaLabel> = [
        .aim, .facebook, .gaduGadu, .googleTalk, .icq, .jabber, .msn, .qq, .skype, .yahoo,
    ]

    private static let serviceMap: [String: String] = [
        "aim": CNInstantMessageServiceAIM,
        "facebook": CNInstantMessageServiceFacebook,
        "flickr": CNSocialProfileServiceFlickr,
        "gameCenter": CNSocialProfileServiceGameCenter,
        "gaduGadu": CNInstantMessageServiceGaduGadu,
        "googleTalk": CNInstantMessageServiceGoogleTalk,
        "icq": CNInstantMessageServiceICQ,
        "jabber": CNInstantMessageServiceJabber,
        "linkedIn": CNSocialProfileServiceLinkedIn,
        "msn": CNInstantMessageServiceMSN,
        "mySpace": CNSocialProfileServiceMySpace,
        "qq": CNInstantMessageServiceQQ,
        "sinaWeibo": CNSocialProfileServiceSinaWeibo,
        "skype": CNInstantMessageServiceSkype,
        "tencentWeibo": CNSocialProfileServiceTencentWeibo,
        "twitter": CNSocialProfileServiceTwitter,
        "yahoo": CNInstantMessageServiceYahoo,
        "yelp": CNSocialProfileServiceYelp,
    ]

    static func fromService(_ service: String, customLabel: String?) -> Label<SocialMediaLabel> {
        for (key, value) in serviceMap where value == service {
            return Label(label: SocialMediaLabel(rawValue: key) ?? .other)
        }
        return Label(label: .custom, customLabel: customLabel ?? service)
    }

    func toService() -> String? {
        return Self.serviceMap[rawValue]
    }

    var isInstantMessage: Bool {
        return Self.instantMessageLabels.contains(self)
    }
}
